src/execution_plan/optimizations/utilize_indices.c

```diff
@@ -19,136 +23,411 @@ AST_Operator _reverseOp(AST_Operator op) {
+			}
 		}
 
 		// Advance to the next operation.
 		current = current->parent;
 	}
+
+	return filters;
 }
 
-// Populate scanOps array with execution plan scan operations.
-void _locateScanOp(OpBase *root, NodeByLabelScan ***scanOps) {
+/* Reduce filter into a range object
+ * Return true if filter was reduce, false otherwise. */
+void _predicateTreeToRange(const FT_FilterNode *tree, rax *string_ranges, rax *numeric_ranges) {
+	// Simple predicate trees are used to build up a range object.
+	assert(tree->pred.lhs->operand.type == AR_EXP_VARIADIC &&
+		   tree->pred.rhs->operand.type == AR_EXP_CONSTANT);
 
-	// Is this a scan operation?
-	if(root->type == OPType_NODE_BY_LABEL_SCAN) {
-		*scanOps = array_append(*scanOps, (NodeByLabelScan *)root);
-	}
+	int op = tree->pred.op;
+	SIValue c = tree->pred.rhs->operand.constant;
+	const char *prop = tree->pred.lhs->operand.variadic.entity_prop;
+	StringRange *sr = raxFind(string_ranges, (unsigned char *)prop, strlen(prop));
+	NumericRange *nr = raxFind(numeric_ranges, (unsigned char *)prop, strlen(prop));
 
-	// Continue scanning.
-	for(int i = 0; i < root->childCount; i++) {
-		_locateScanOp(root->children[i], scanOps);
+	// Get or create range object for alias.prop.
+	if(SI_TYPE(c) & SI_NUMERIC || SI_TYPE(c) == T_BOOL) {
+		// Create if doesn't exists.
+		if(nr == raxNotFound) {
+			nr = NumericRange_New();
+			raxTryInsert(numeric_ranges, (unsigned char *)prop, strlen(prop), nr, NULL);
+		}
+		NumericRange_TightenRange(nr, op, SI_GET_NUMERIC(c));
+	} else if(SI_TYPE(c) & SI_STRING) {
+		// Create if doesn't exists.
+		if(sr == raxNotFound) {
+			sr = StringRange_New();
+			raxTryInsert(string_ranges, (unsigned char *)prop, strlen(prop), sr, NULL);
+		}
+		StringRange_TightenRange(sr, op, c.stringval);
+	} else {
+		assert(false);
 	}
 }
 
-void utilizeIndices(GraphContext *gc, ExecutionPlan *plan) {
-	// Return immediately if the graph has no indices
-	if(!GraphContext_HasIndices(gc)) return;
+/* Try to replace given Label Scan operation and a set of Filter operations with
+ * a single Index Scan operation. */
+void reduce_scan_op(ExecutionPlan *plan, NodeByLabelScan *scan) {
+	RSQNode *root = NULL;
+	uint rsqnode_count = 0;
 
-	// Collect all label scans
-	NodeByLabelScan **scanOps = array_new(NodeByLabelScan *, 0);
-	_locateScanOp(plan->root, &scanOps);
+	// Make sure there's an index for scanned label.
+	const char *label = scan->node->label;
+	GraphContext *gc = GraphContext_GetFromTLS();
+	Index *idx = GraphContext_GetIndex(gc, label, NULL, IDX_EXACT_MATCH);
+	if(idx == NULL) return;
 
-	// Collect all filters on scanned entities
-	NodeByLabelScan *scanOp;
-	OpBase **filterOps = array_new(OpBase *, 0);
-	FT_FilterNode *ft;
-	const char *label;
+	/* Reduce filters into ranges.
+	 * we differentiate between between numeric filters
+	 * and string filters. */
+	rax *string_ranges = raxNew();
+	rax *numeric_ranges = raxNew();
+	RSQNode **rsqnodes = array_new(RSQNode *, 1);
```

Comment:
Let's NULL-initialize these variables and move the allocs to 316, since we don't have a strong reason to believe that we'll find valid filters.

Reply:
👍